pkg/profile/artifact.go

```diff
@@ -7,54 +7,180 @@ import (
+		if err := p.client.Create(ctx, helmRes); err != nil {
 			return fmt.Errorf("failed to create HelmRelease resource: %w", err)
 		}
-	case "Kustomize":
-		if err := p.createKustomization(ctx); err != nil {
+	} else {
+		p.log.Info("creating Kustomization", "resource", kustomizeRes.ObjectMeta.Name)
+		if err := p.client.Create(ctx, kustomizeRes); err != nil {
 			return fmt.Errorf("failed to create Kustomization resource: %w", err)
 		}
-	default:
-		return fmt.Errorf("artifact kind %q not recognized", kind)
 	}
-
 	p.log.Info("all artifacts created")
 	return nil
 }
 
+// Checks if the artifacts exists and returns any ready!=true conditions on the artifacts.
+func (p *Profile) ArtifactStatus(ctx context.Context) (Status, error) {
+	resourcesExist, gitRes, helmRes, kustomizeRes, err := p.getResources(ctx)
+	if err != nil {
+		return Status{}, err
+	}
+
+	return Status{
+		ResourcesExist:     resourcesExist,
+		NotReadyConditions: p.checkResourcesReady(gitRes, helmRes, kustomizeRes),
+	}, nil
+}
+
+func (p *Profile) checkResourcesReady(gitRes *sourcev1.GitRepository, helmRes *helmv2.HelmRelease, kustomizeRes *kustomizev1.Kustomization) []metav1.Condition {
+	var notReadyConditions []metav1.Condition
+	if gitRes != nil {
+		if condition := getNotReadyCondition(gitRes.Status.Conditions); condition != (metav1.Condition{}) {
+			notReadyConditions = append(notReadyConditions, condition)
+		}
+	}
+
+	if helmRes != nil {
+		if condition := getNotReadyCondition(helmRes.Status.Conditions); condition != (metav1.Condition{}) {
+			notReadyConditions = append(notReadyConditions, condition)
+		}
+	}
+
+	if kustomizeRes != nil {
+		if condition := getNotReadyCondition(kustomizeRes.Status.Conditions); condition != (metav1.Condition{}) {
+			notReadyConditions = append(notReadyConditions, condition)
+		}
+	}
+	return notReadyConditions
+}
+
+func (p *Profile) getResources(ctx context.Context) (bool, *sourcev1.GitRepository, *helmv2.HelmRelease, *kustomizev1.Kustomization, error) {
+	gitRes, helmRes, kustomizeRes, err := p.makeArtifacts()
+	if err != nil {
+		return false, nil, nil, nil, err
+	}
+	gitResExists := true
+	helmResExists := true
+	kustomizeResExists := true
+	if gitRes != nil {
+		gitResExists, err = p.getResourceIfExists(ctx, gitRes)
+		if err != nil {
+			return false, nil, nil, nil, err
+		}
+	}
+
+	if helmRes != nil {
+		helmResExists, err = p.getResourceIfExists(ctx, helmRes)
+		if err != nil {
+			return false, nil, nil, nil, err
+		}
+	}
+
+	if kustomizeRes != nil {
+		kustomizeResExists, err = p.getResourceIfExists(ctx, kustomizeRes)
+		if err != nil {
+			return false, nil, nil, nil, err
+		}
+	}
+	return gitResExists && helmResExists && kustomizeResExists, gitRes, helmRes, kustomizeRes, nil
+}
+
+func getNotReadyCondition(conditions []metav1.Condition) metav1.Condition {
+	for _, condition := range conditions {
+		if condition.Type == "Ready" && string(condition.Status) != "True" {
+			return condition
+		}
+	}
+	return metav1.Condition{}
```

Comment:
What do you think about instead of comparing to an empty struct, return a `ready` boolean? You do that in other places, I think it would be nice and consistent to do that here too?

Reply:
I did have that originally, but it refactored it out as it looked ugly haha, I'm quite torn on which looks better. I sort of think this approach is equiavlent to returning `error` and comparing `err!=nil`. WDYT?

Reply:
Heh, yeah. I can see that. :)
I usually avoid the structure compare because it works only as long as all the fields are comparable in the struct and then if you add another field which isn't, like a slice, then this part of the code will break. Which is fine I guess, but then you'd have to update the code to also compare slices and slice values and that's just a drag really. :)

So in the long term it's usually more convenient to just return a yay or nay if possible.

I'm not trying to convince you, I just wanted to tell you my reasoning behind it. :)

Reply:
mhmm thats fair, in this case we are relying on a pretty standard struct that shouldn't realistically ever be changing in format (this is a core part of kubernetes api), I'm inclined to leave it as it is

Reply:
Sure, I'm fine with that. :) 👍